File: src/main/java/com/morph/engine/input/Keyboard.java
package com.morph.engine.input;

import static com.morph.engine.input.Keyboard.StdKeyAction.PRESS;
import static com.morph.engine.input.Keyboard.StdKeyAction.RELEASE;
import static org.lwjgl.glfw.GLFW.*;

import com.morph.engine.util.Feed;
import com.morph.engine.util.Listener;
import io.reactivex.Maybe;
import io.reactivex.Observable;

import java.util.Arrays;

public class Keyboard {
	private static Feed<StdKeyEvent> keyEventFeed = new Feed<>();

	// PRESS, REPEAT, RELEASE
	private static Observable<StdKeyEvent> standardKeyEvents = Observable.create(keyEventFeed::emit);

	// UP, DOWN
	private static Observable<BinKeyEvent> binaryKeyEvents = Observable.create(emitter -> {
		Listener<StdKeyEvent> listener = new Listener<StdKeyEvent>() {
			@Override
			public void onNext(StdKeyEvent stdKeyEvent) {
				switch(stdKeyEvent.action) {
					case PRESS:
						emitter.onNext(new BinKeyEvent(BinKeyAction.DOWN, stdKeyEvent.getKey(), stdKeyEvent.getMods()));
						break;
					case RELEASE:
						emitter.onNext(new BinKeyEvent(BinKeyAction.UP, stdKeyEvent.getKey(), stdKeyEvent.getMods()));
						break;
					case REPEAT:
						break;
				}
			}

			@Override
			public void onError(Throwable t) {
				emitter.onError(t);
			}

			@Override
			public void onComplete() {
				emitter.onComplete();
			}
		};
		keyEventFeed.register(listener);
	});

	public enum StdKeyAction {
		PRESS, REPEAT, RELEASE
	}

	public enum BinKeyAction {
		UP, DOWN
	}

	public static class GenericKeyEvent {
		private int key;
		private int mods;

		public GenericKeyEvent(int key, int mods) {
			this.key = key;
			this.mods = mods;
		}

		public int getKey() {
			return key;
		}

		public int getMods() {
			return mods;
		}

		public boolean hasMod(int modCheck) {
			return (mods & modCheck) != 0;
		}
	}

	public static class StdKeyEvent extends GenericKeyEvent {
		private StdKeyAction action;

		public StdKeyEvent(StdKeyAction action, int key, int mods) {
			super(key, mods);
			this.action = action;
		}

		public StdKeyAction getAction() {
			return action;
		}
	}

	public static class BinKeyEvent extends GenericKeyEvent {
		private BinKeyAction action;

		public BinKeyEvent(BinKeyAction action, int key, int mods) {
			super(key, mods);
			this.action = action;
		}

		public BinKeyAction getAction() {
			return action;
		}
	}

	public static void handleKeyEvent(long window, int key, int scancode, int action, int mods) {
		keyEventFeed.onNext(new StdKeyEvent(getKeyAction(action), key, mods));
	}

	private static StdKeyAction getKeyAction(int action) {
		switch (action) {
			case GLFW_PRESS:
				return PRESS;
			case GLFW_REPEAT:
				return StdKeyAction.REPEAT;
			case GLFW_RELEASE:
				return RELEASE;
			default:
				return null;
		}
	}

	public static boolean queryUpDown(int key, BinKeyAction action) {
		Maybe<BinKeyEvent> e = binaryKeyEvents.filter(event -> event.getKey() == key).lastElement();
		return !e.isEmpty().blockingGet() && e.blockingGet().action == action;
	}

	public static boolean queryUpDownWithMods(int key, BinKeyAction action, int... mods) {
		Maybe<BinKeyEvent> e = binaryKeyEvents.filter(event -> event.getKey() == key).lastElement();
		return !e.isEmpty().blockingGet() && e.blockingGet().action == action && Arrays.stream(mods).allMatch(e.blockingGet()::hasMod);
	}

	public static Observable<StdKeyEvent> getStandardKeyEvents() {
		return standardKeyEvents;
	}
	public static Observable<BinKeyEvent> getBinaryKeyEvents() {
		return binaryKeyEvents;
	}
}

File: src/main/kotlin/com/morph/engine/input/Keyboard.kt
package com.morph.engine.input

import org.lwjgl.glfw.GLFW.*

object Keyboard {
    // PRESS, REPEAT, RELEASE
    private val keyEvents: MutableList<StdKeyEvent> = mutableListOf()
    @JvmStatic val standardKeyEvents: MutableList<StdKeyEvent> = keyEvents

    private val keyDowns: Array<Pair<Boolean, Int>> = Array(GLFW_KEY_LAST - GLFW_KEY_SPACE + 1) { false to 0 }

    // UP, DOWN
    @JvmStatic val binaryKeyEvents: List<BinKeyEvent>
        get() = keyDowns
            .mapIndexed { index, (keyDown, mods) ->
                BinKeyEvent(if (keyDown) KeyDown else KeyUp, index + GLFW_KEY_SPACE, mods)
            }

    fun handleKeyEvent(window: Long, key: Int, scancode: Int, action: Int, mods: Int) {
        if (key - GLFW_KEY_SPACE >= 0) {
            keyEvents.add(StdKeyEvent(getKeyAction(action), key, mods))
            if (action == GLFW_PRESS) keyDowns[key - GLFW_KEY_SPACE] = true to mods
            if (action == GLFW_RELEASE) keyDowns[key - GLFW_KEY_SPACE] = false to 0
        }
    }

    private fun getKeyAction(action: Int): StdKeyAction = when (action) {
        GLFW_PRESS -> KeyPress
        GLFW_REPEAT -> KeyRepeat
        GLFW_RELEASE -> KeyRelease
        else -> throw IllegalArgumentException("Argument is not a valid input action")
    }

    fun queryUpDown(key: Int, action: BinKeyAction): Boolean {
        val e = binaryKeyEvents.last { event -> event.key == key }
        return e.action == action
    }

    fun queryUpDownWithMods(key: Int, action: BinKeyAction, vararg mods: Int): Boolean {
        val e = binaryKeyEvents.last { event -> event.key == key }
        return e.action == action && mods.all { e.hasMod(it) }
    }

    fun clear() {
        keyEvents.clear()
        standardKeyEvents.clear()
    }

    fun update() {
    }
}


sealed class KeyAction
sealed class StdKeyAction : KeyAction()
object KeyPress : StdKeyAction()
object KeyRepeat : StdKeyAction()
object KeyRelease : StdKeyAction()

sealed class BinKeyAction : KeyAction()
object KeyUp : BinKeyAction()
object KeyDown : BinKeyAction()

enum class KeyActions(val action: KeyAction) {
    PRESS(KeyPress), REPEAT(KeyRepeat), RELEASE(KeyRelease),
    UP(KeyUp), DOWN(KeyDown)
}

sealed class GenericKeyEvent(val key: Int, val mods: Int) {
    fun hasMod(modCheck: Int): Boolean {
        return mods and modCheck != 0
    }
}
class StdKeyEvent(val action: StdKeyAction, key: Int, mods: Int) : GenericKeyEvent(key, mods)
class BinKeyEvent(val action: BinKeyAction, key: Int, mods: Int) : GenericKeyEvent(key, mods)
